File: app/src/main/java/com/example/steaminvestmentmanager/utilclasses/CurrencyData.java
package com.example.steaminvestmentmanager.utilclasses;

import com.example.steaminvestmentmanager.MainActivity;
import com.google.gson.Gson;
import java.text.DecimalFormat;


public class CurrencyData {
    private static int currency = 0;
    private static final Gson gson = new Gson();
    private static final String[] currencies = new String[]{"$", "$", "£", "€", "", "руб."};

    public static String getCurrencyChar() {
        try {
            return currencies[currency];
        } catch (ArrayIndexOutOfBoundsException e) {
            return "$";
        }
    }

    public static int getCurrency() {
        return currency;
    }

    public static String[] getCurrencyArray() {
        return new String[] {"$", "£", "€", "руб."};
    }

    public static int getCurrencyArrayPosition() {
        for (int i = 0; i < getCurrencyArray().length; i++) {
            if (getCurrencyArray()[i].equals(getCurrencyChar())) return i;
        }
        return -1;
    }

    public static int getCurrencyFromChar(String currencyString) {
        for (int i = 0; i < currencies.length; i++) {
            if (currencyString.equals(currencies[i])) return i;
        }
        return 0;
    }

    public static void setInitialCurrency(int currency) {
        CurrencyData.currency = currency;
    }

    public static void setCurrency(int currency) {
        CurrencyData.currency = currency;
        SteamItem[] items = MainActivity.getSteamItems();
        if (items == null) return;
        for (SteamItem item:
             items) {
            if (item == null) continue;
            float ration = getCurrencyToCurrency(item.getCurrentCurrency(), currency);
            item.setBuyingPrice(Float.parseFloat(new DecimalFormat("#.##").format(item.getBuyingPrice() / ration).replace(',', '.')));
            item.setCurrentPrice(Float.parseFloat(new DecimalFormat("#.##").format(item.getCurrentPrice() / ration).replace(',', '.')));
            item.setCurrentCurrency(currency);
        }
        ItemListUpdater.updateSteamItemAdapter();
    }

    public static float transformPriceToNumber(String price) {
        price = price.replaceAll("pуб.", "").replaceAll("\\$", "").replaceAll(",", ".").replaceAll("€", "").replaceAll("£", "").replaceAll("CHF", "");
        return Float.parseFloat(price);
    }

    private static final String firstPartOfURL = "https://steamcommunity.com/market/priceoverview/?currency=", secondPartOfURL = "&country=us&appid=440&market_hash_name=Mann%20Co.%20Supply%20Crate%20Key&format=json";

    public static float getCurrencyToCurrency(int firstCurrency, int secondCurrency){
        if (firstCurrency == secondCurrency) return 1.0f;
        String convURL1 = firstPartOfURL + firstCurrency + secondPartOfURL;
        String jsonSteamItem1 = new DownloadingPageHtmlCode(convURL1).call();
        PriceoverviewSteamItem item1 = gson.fromJson(jsonSteamItem1, PriceoverviewSteamItem.class);

        String convURL2 = firstPartOfURL + secondCurrency + secondPartOfURL;
        String jsonSteamItem2 = new DownloadingPageHtmlCode(convURL2).call();
        PriceoverviewSteamItem item2 = gson.fromJson(jsonSteamItem2, PriceoverviewSteamItem.class);
        if (item1 == null || item2 == null) return -1;
        return transformPriceToNumber(item1.getCurrentPrice()) / transformPriceToNumber(item2.getCurrentPrice());
    }

}

File: app/src/main/java/com/example/steaminvestmentmanager/utilclasses/CurrencyData.kt
package com.example.steaminvestmentmanager.utilclasses

import android.util.Log
import com.example.steaminvestmentmanager.MainActivity.Companion.getSteamItems
import com.google.gson.Gson
import java.text.DecimalFormat

object CurrencyData {
    private var currency = 0
    private val gson = Gson()
    private val currencies = arrayOf("$", "$", "£", "€", "CHF", "₽")

    var exchangeRatios = HashMap<Int, Float>()

    init {
        exchangeRatios[0] = 1.0f
    }

    fun updateExchangeRatios() {
        for (i in currencyArray) {
            if (!i.equals("$")) {
                val iPos = getCurrencyFromChar(i)
                Log.d("SteamIvan", "$i")
                exchangeRatios[iPos] = getCurrencyToCurrency(0, iPos)
            }
        }
    }

    val currencyChar: String
        get() = try {
            currencies[currency]
        } catch (e: ArrayIndexOutOfBoundsException) {
            "$"
        }

    fun getCurrency(): Int {
        return currency
    }

    val currencyArray: Array<String?>
        get() = arrayOf("$", "£", "€", "₽")
    val currencyArrayPosition: Int
        get() {
            for (i in currencyArray.indices) {
                if (currencyArray[i] == currencyChar) return i
            }
            return -1
        }

    fun getCurrencyFromChar(currencyString: String?): Int {
        for (i in currencies.indices) {
            if (currencyString == currencies[i]) return i
        }
        return 0
    }

    fun setInitialCurrency(currency: Int) {
        CurrencyData.currency = currency
    }

    fun setCurrency(currency: Int) {
        CurrencyData.currency = currency
        val items = getSteamItems()
        for (item in items) {
            val ratio = exchangeRatios[item.currentCurrency]?.div(exchangeRatios[currency]!!)
            item.buyingPrice = DecimalFormat("#.##").format((item.buyingPrice * ratio!!).toDouble()).replace(',', '.').toFloat()
            item.currentPrice = DecimalFormat("#.##").format((item.currentPrice * ratio).toDouble()).replace(',', '.').toFloat()
            item.currentCurrency = currency
        }
        ItemListUpdater.updateSteamItemAdapter()
    }

    fun transformPriceToNumber(price: String?): Float {
        return price!!.replace("pуб.".toRegex(), "").replace("\\$".toRegex(), "").replace(",".toRegex(), ".").replace("€".toRegex(), "").replace("£".toRegex(), "").replace("CHF".toRegex(), "").toFloat()
    }

    private const val firstPartOfURL = "https://steamcommunity.com/market/priceoverview/?currency="
    private const val secondPartOfURL = "&country=us&appid=440&market_hash_name=Mann%20Co.%20Supply%20Crate%20Key&format=json"
    fun getCurrencyToCurrency(firstCurrency: Int, secondCurrency: Int): Float {
        if (firstCurrency == secondCurrency) return 1.0f
        val convURL1 = firstPartOfURL + firstCurrency + secondPartOfURL
        val jsonSteamItem1 = DownloadingPageHtmlCode(convURL1).call()
        val item1 = gson.fromJson(jsonSteamItem1, PriceoverviewSteamItem::class.java)

        val convURL2 = firstPartOfURL + secondCurrency + secondPartOfURL
        val jsonSteamItem2 = DownloadingPageHtmlCode(convURL2).call()
        val item2 = gson.fromJson(jsonSteamItem2, PriceoverviewSteamItem::class.java)
        return if (item1 == null || item2 == null) (-1.0f) else transformPriceToNumber(item1.lowest_price) / transformPriceToNumber(item2.lowest_price)
    }


}